From 6365adb17759eb19a2b51e8e4f8c21e9a4ed6ef6 Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Thu, 6 Jul 2023 14:31:18 +0200
Subject: [PATCH] WLANRTSYS-56718: IOP fixes implementation

- STA: ignore deauth flag causes roaming failure

    In case AP sends disassociate to sta wpa_supplicant generates a deauthenticate
    event to clear driver state and also turns on the ignore_next_local_deauth flag
    expecting to get a deauth event for a locally generated deauth.
    However, no deauth or disconnect events are received, since sta is already
    marked as disconnected.
    AP will connect to AP2 (the new AP) but if connection to AP2 is lost deauth
    event to wpa_supplicant will be ignored since the ignore_next_local_deauth
    was never set to 0. So wpa_supplicant is unaware that station is disconnected
    and does not initiate scan.
    Fix: Set ignore_next_local_deauth flag to 0 upon connection.
---
 src/drivers/driver_nl80211_event.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 7082c1d02..ea60f4b86 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -311,6 +311,20 @@ static void mlme_event_assoc(struct wpa_driver_nl80211_data *drv,
 	}
 
 	drv->associated = 1;
+
+	if (drv->ignore_next_local_deauth == 1){
+		/*
+		 * In case the flag to ignore deauth event triggered by
+		 * local deauthentication was not reset (because kernel did not
+		 * actually send the deauth event) reset ignore_next_local_deauth
+		 * flag
+		 */
+		wpa_printf(MSG_DEBUG,
+			"nl80211: Resetting flag that prevents raising deauth event"
+			" triggered by local deauth");
+			drv->ignore_next_local_deauth = 0;
+		}
+
 	os_memcpy(drv->bssid, mgmt->sa, ETH_ALEN);
 	os_memcpy(drv->prev_bssid, mgmt->sa, ETH_ALEN);
 
-- 
2.43.0

